src/librustc_middle/traits/structural_impls.rs

```diff
@@ -232,11 +233,7 @@ impl<'a, 'tcx> Lift<'tcx> for traits::DerivedObligationCause<'a> {
 impl<'a, 'tcx> Lift<'tcx> for traits::ObligationCause<'a> {
     type Lifted = traits::ObligationCause<'tcx>;
     fn lift_to_tcx(&self, tcx: TyCtxt<'tcx>) -> Option<Self::Lifted> {
-        tcx.lift(&self.code).map(|code| traits::ObligationCause {
-            span: self.span,
-            body_id: self.body_id,
-            code,
-        })
+        tcx.lift(&self.code).map(|code| traits::ObligationCause::new(self.span, self.body_id, code))
```

Comment:
It feels like there should be a way to clone the `ObligationCause` here when the `ObligationCauseCode` is one of the variants without `'tcx` instead of allocating a new `Lrc` for it. I suppose this applies equally to the impl for `DerivedObligationCause` above, so maybe I'm wrong? At least leave a note here about the possibility to optimize further.

Reply:
Afaict lift is pretty much irrelevant. Compiling parts of `stage 2` had 0 calls to lift

Reply:
So I don't want to transmute lifetimes here